Dockerfile

```diff
@@ -0,0 +1,42 @@
+################################################################################################################ builder
+
+FROM maven:3-jdk-8 as builder
+WORKDIR /build
+
+# this block will make docker cache the dependencies for subsequent builds of the same image.
+COPY pom.xml .
+COPY subsonic-rest-api/pom.xml ./subsonic-rest-api/pom.xml
+COPY integration-test/pom.xml ./integration-test/pom.xml
+COPY airsonic-sonos-api/pom.xml ./airsonic-sonos-api/pom.xml
+COPY airsonic-main/pom.xml ./airsonic-main/pom.xml
+RUN mvn de.qaware.maven:go-offline-maven-plugin:resolve-dependencies
+
+COPY . .
+RUN mvn package
```

Comment:
maybe change this to verify to ensure integration-tests (if any) get run?

Reply:
I would even argue to use `-DskipTests=true` here and leave the process of running integration tests and tagging a release to the CI pipeline. Running the integration tests here would result in a DinD scenario.

Reply:
oo, good point.

so then help me understand why a multistage build is needed. Is it just for stand-alone docker builds especially where you don't have/want mvn on the host?

Let's examine the build scenario:
- CI runs `mvn verify`
  - after `package`, the `pre-integration-test` phase needs and builds the docker container,   which, in turn, runs `mvn package` internally and compiles and packages again (even though an artifact already exists outside and you can just copy it)
  - the verify is done against the newly compiled/packaged artifact sitting inside the container

Reply:
> "Is it just for stand-alone docker builds especially where you don't have/want mvn on the host?" 

In Short: yes. In that particular case for easy integration with Docker Hub. 

From personal experience / opinion:
Expecting / imposing dependencies to specific toolchains on the CI/CD build agent is prone to errors and the root of many failed builds. I like approaches based purely on docker such as Drone.io. So basically boiling down to just one and leaving the responsibility to use the right JDK or maven version required for building to the developers and not bother CI/CD operators with tasks to maintain possibly exotic toolchains. 

That said: my perspective is perhaps biased and affected by my daytime job. Take that opinion  with a grain of salt.

Reply:
> Expecting / imposing dependencies to specific toolchains on the CI/CD build agent is prone to errors and the root of many failed builds.

I'd argue this is the sole purpose of using Docker and any CI/CD. This ensures that wherever Docker is running, it will work because it has the required toolchain. Like you said,

> not bother CI/CD operators with tasks to maintain possibly exotic toolchains

I'd agree here, but it's important that we make sure that we have all the required dependencies within the package and Docker image.

Reply:
> In Short: yes. In that particular case for easy integration with Docker Hub.

I see. Without running any tests?

Not arguing the utility of a standalone Dockerfile (which certainly has value), but I think it'd be nicer to have a CI/CD push non-failed builds only to Docker Hub.

But of course, that's just my subjective opinion.

Reply:
> But of course, that's just my subjective opinion.

I guess mine too 🤣

Reply:
not sure we are on the same page. I do not intend to remove the unit tests which are run implicitly by the `package` goal. 

Skipping the tests does only make sense if the release (tagging) is triggered by something like `mvn release` which must run tests and ideally also the integration tests before tagging a release. This is currently out of this PR`s scope. 

May I resolve this thread or am I missing a point here?

Reply:
yes you may resolve this.

I was just pointing out that I don't think this could be part of the release process, it's an independent alternative way to build.

> I do not intend to remove the unit tests which are run implicitly by the package goal.

Why not? Since this is not part of any release, why have unit tests running? Or run unit but skip the integration-tests?

The release process that'll build it's own image, won't need this Dockerfile, because it already has access to the artifact (and doesn't need to compile it again). Its simplified Dockerfile merely copies the artifact into a base container.